### PR TITLE
Simplify minifier testing by incorporating fault injection in prod code

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100447
* #100324
* #100416
* #100374
* #100387
* __->__ #100357
* #100354

Previously, minifier testing injected faults by injecting extra code
into the repro scripts, and then ensuring this code got propagated to
all subsequent subprocess calls.  This was not only quite complicated,
but also induced a big slowdown on the minifier, because to inject the
faults, you had to import torch._inductor, which would cause the
compilation threads to immediately get initialized before you even got
to do anything else in the repro script.

This new approach fixes this problem by incorporating the fault
injection into "prod" code.  Essentially, for inductor fault injection
we introduce some new config flags that let you "configure" Inductor to
be buggy; for Dynamo fault injection we just permanently keep the buggy
testing backends registered.  This is MUCH simpler: we only have to
propagate the buggy config (which is something we're already doing),
and it saves the minifier scripts from having to immediately initialize
inductor on entry.

Also, I enable the test for Triton runtime errors, now that tl.assert_device is here.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/dynamo/test_minifier.py [hunk 1]
```diff
@@ -142,64 +86,48 @@ def inner(x):
...
+            "cuda", 2, "relu_runtime_error_TESTING_ONLY"
+        )
 
     @requires_cuda()
     def test_after_dynamo_cuda_accuracy_backend_passes(self):
-        self._test_after_dynamo_backend_passes("cuda", 4, RELU_ACCURACY_ERROR_BACKEND)
-
-    # Ensure that generated code with a custom backends generates a runnable minifier
-    # launcher script that results in a RuntimeError
-    def test_after_dynamo_custom_backend(self):
```
Comment (ezyang) [inline test/dynamo/test_minifier.py:173]: I didn't really understand what this test was testing, that the compile time error backend isn't already testing.

File: test/dynamo/test_minifier.py [hunk 2]
```diff
@@ -326,54 +253,6 @@ def test_dynamo_config_serialization(self):
         proc, _ = self._run_test_code(run_code)
         self.assertEqual(proc.returncode, 0)
 
-    # Test that launched minifier processes have the same config as
-    # the original process.
-    def _test_after_dynamo_with_modified_config(
```
Comment (ezyang) [inline test/dynamo/test_minifier.py:331]: This test is now pointless; if config is not propagated, the bugs won't propagate (as they are propagated by config!)

File: test/inductor/minifier_smoke.py [hunk 3]
```diff
@@ -1,60 +0,0 @@
-# Owner(s): ["module: inductor"]
-import os
-
-os.environ["TORCHDYNAMO_REPRO_AFTER"] = "dynamo"
```
Comment (ezyang) [inline test/inductor/minifier_smoke.py:4]: This file looked bitrotten, and didn't seem to be referenced from anywhere else

Comment (mlazos) [reply test/inductor/minifier_smoke.py:4]: This was intended to be a manual test like the inductor smoke test (it's referenced in the minifier runbook). It uses a similar API as the other minifier tests so I didn't have it run automatically.

Comment (ezyang) [reply test/inductor/minifier_smoke.py:4]: For reference, this appears to be the runbook https://docs.google.com/document/d/18L9e7bZSBpJ7gGbwlUV13LasmjiEX2lree2pl-SdbCU/edit

Comment (ezyang) [reply test/inductor/minifier_smoke.py:4]: restored the file

File: torch/_dynamo/repro/after_aot.py [hunk 4]
```diff
@@ -395,9 +394,7 @@ def symint(self, name, val) -> None:
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
 
 
-def generate_compiler_repro_string(
-    gm, args, *, stable_output=False, save_dir=None, patch_code=None
-):
+def generate_compiler_repro_string(gm, args, *, stable_output=False, save_dir=None):
```
Comment (bdhirsh) [inline torch/_dynamo/repro/after_aot.py:397]: I remember briefly reading this and wondering what `patch_code` was meant for 🤔

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -198,7 +198,21 @@ def libdevice_sqrt(x):
...
     @staticmethod
     def relu(x):
-        return ops.maximum("0", x)
+        bug = config.triton.inject_relu_bug_TESTING_ONLY
+        if bug == "compile_error":
+            return "compile error!"
+        elif bug == "runtime_error":
+            # NB: this only triggers runtime error as long as input
+            # is not all zero
+            return f'triton_helpers.device_assert_then({x} == 0, "injected assert fail", {x})'
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_inductor/codegen/triton.py:207]: I tried passing False here but Triton choked on it. Maybe there is some other way to spell this, suggestions from Triton experts would be useful. (maybe @peterbell10 or @lezcano knows?)

[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:207]: `device_assert` takes a tensor, so perhaps `tl.zeros` works?

**Review by voznesenskym — APPROVED** [review]

File: torch/_inductor/codegen/triton.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:207]: Now `device_assert(False,...)` should work on triton once we update the pin after https://github.com/openai/triton/pull/1641

**Pull request closed without merging**